onnxruntime/contrib_ops/cpu/quantization/qlinear_softmax.cc

```diff
@@ -0,0 +1,358 @@
+                                   size_t reduce_len, bool is_signed) {
+  const double qscale =
+      fmin(static_cast<double>(UINT32_MAX) / static_cast<double>(reduce_len), static_cast<double>(0x7fffff));
+  for (int32_t i = 0; i < 256; i++) {
+    double scaled_exp_xi = qscale * exp(static_cast<double>(i - 255) * static_cast<double>(x_scale));
+    // we can't get the real max value of input tensor here, so we just assume 255.
+    // in the function of `QlinearSoftmaxCPU`,
+    // all numbers will have a shift (255-max_value) if its max value is not 255
+    //
+    // if is_signed index = [1 2 3 ......126 127 -128 -127 ..... -3 -2 -1]
+    // else [0 1 2 3 4 ..... 256]
+    uint8_t index = static_cast<uint8_t>(is_signed ? i - 128 : i);
+    table[index] = static_cast<uint32_t>(lrint(scaled_exp_xi));
+  }
+}
+
+void BuildLookupTableIfFixed(const OpKernelInfo& info, std::vector<uint32_t>& fixed_lookup_table,
+                             size_t reduce_len, bool is_signed) {
+  const Tensor* tensor_x_scale = nullptr;
+
+  bool get_x_scale = info.TryGetConstantInput(1, &tensor_x_scale);
+  ORT_ENFORCE(tensor_x_scale == nullptr || IsScalarOr1ElementVector(tensor_x_scale),
+              "QlinearBuildLookupTable : input X_scale must be a scalar or 1D tensor of size 1");
+  bool is_fixed_parameters = get_x_scale;
+
+  if (is_fixed_parameters) {
+    fixed_lookup_table.resize(256);
+    const float X_scale = *(tensor_x_scale->Data<float>());
+    QlinearBuildLookupTableUint32(fixed_lookup_table, X_scale, reduce_len, is_signed);
+  }
+}
+}  // namespace
+
+
+QLinearSoftmax::QLinearSoftmax(const OpKernelInfo& info)
+    : OpKernel(info) {
+  const auto& node = info.node();
+  auto input_defs = node.InputDefs();
+  auto input_type = input_defs[0]->TypeAsProto()->tensor_type().elem_type();
+  is_signed_ = (input_type == ONNX_NAMESPACE::TensorProto_DataType_INT8);
+  const auto* x_shape = input_defs[0]->Shape();
+  if (x_shape == nullptr || x_shape->dim_size() == 0) {
+    return;
+  }
+  int rank = x_shape->dim_size();
+
+  int64_t opset = -1;
+  Status status = info.GetAttr<int64_t>("opset", &opset);
+  ORT_ENFORCE(status.IsOK(), "opset must be existed in attributes of QlinearSoftmax");
+  opset_ = gsl::narrow_cast<int>(opset);
+
+  int64_t axis = -1;
+  status = info.GetAttr<int64_t>("axis", &axis);
+  if (status.IsOK()) {
+    axis_ = gsl::narrow_cast<int>(axis);
+  } else {
+    if (opset_ < OPSET13) {
+      axis_ = 1;  // opset-12 and below, the default axis value is 1
+    } else {
+      axis_ = -1;  // opset-13, the default axis value is -1
+    }
+  }
+
+  if (axis_ < 0) {
+    axis_ = static_cast<int>(HandleNegativeAxis(axis_, int64_t(rank)));
+  }
+  auto input_shape = utils::GetTensorShapeFromTensorShapeProto(*x_shape);
+  int64_t reduce_size = input_shape[axis_];
+  if (opset_ < OPSET13) {
+    reduce_size = input_shape.SizeFromDimension(axis_);
+  }
+  ORT_ENFORCE(reduce_size > 0, "invalid reduce_size for softmax");
```

Comment:
If there are any dynamic dimensions I believe x_shape could have -1 values in it - which is perfectly legal. 

reduce_size should be -1 if that is the case (from either input_shape[axis_] or SizeFromDimension) and you'll need to calculate the value in each call to Compute.

Reply:
Yes, It's possible to have the performing axis to be dynamic.